# 9주차 과제
목표
- 자바의 예외 처리에 대해 학습하세요.

학습할것 필수
- 자바에서 예외 처리 방법 (try, catch, throw, throws, finally)
- 자바가 제공하는 예외 계층 구조
- Exception과 Error의 차이는?
- RuntimeException과 RE가 아닌 것의 차이는?
- 커스텀한 예외 만드는 방법

## 자바에서 예외 처리 방법 (try, catch, throw, throws, finally)
예외처리란?
- 프로그램 실행시 발생 할 수 있는 예외의 발생에 대비한 코드를 작성하는것
- 프로그램의 비정상 종료를 막고, 정상적인 실행 상태를 유지하는것

try-catch
- 예외를 처리하기 위해서 try-catch문 사용. 구조는 다음과 같음
~~~java
try {
    // 예외가 발생할 가능성이 있는 문장들을 넣는다.
} catch (Exception1 e1){
    // Exception1 이 발생했을 경우, 이를 처리하기 위한 문장을 적는다.
} catch (Exception2 e2){
  // Exception2가 발생했을 경우, 이를 처리하기 위한 문장을 적는다.
}
...
} catch (ExceptionN eN){
  // ExceptionN이 발생했을 경우, 이를 처리하기 위한 문장을 적는다.
}
~~~

try블럭 내에서 예외가 발생한 경우
1. 발생한 예외와 일치하는 catch블럭이 있는지 확인한다.
2. 일치하는 catch 블럭을 찾게되면, 그 catch블럭 내의 문장들을 수행하고 전체 try-catch문을 빠져나가서 그다음 문장을 계속해서 수행한다. 만일 일치하는 catch 블럭을 찾지 못하면 , 예외는 처리되지 못한다

~~~java
public class ExceptionEx {

    public static void main(String[] args) {
        System.out.println(1);
        System.out.println(2);

        try {
            System.out.println(3);
            System.out.println(0/0); // 고의로 ArithmeticException 발생
            System.out.println(4);
        } catch (Exception e) {
            System.out.println(5);
            e.printStackTrace();
        }// try-catch 끝

        System.out.println(6);
        // main 메소드 끝
    }
}

~~~
- 실행결과
~~~
1
2
3
5
6
~~~


try블럭 내에서 예외가 발생하지 않은 경우,
1. catch 블럭을 거치지 않고 전체 try-catch문을 빠져나가서 수행을 계속한다.
~~~java
public class ExceptionEx {

    public static void main(String[] args) {
        System.out.println(1);
        System.out.println(2);

        try {
            System.out.println(3);
            System.out.println(4);
        } catch (Exception e) {
            System.out.println(5);
            e.printStackTrace();
        }// try-catch 끝

        System.out.println(6);
        // main 메소드 끝 
    }
}
~~~
- 실행결과
~~~
1
2
3
4
6
~~~

throw
- 키워드 thow 를 사용하여 프로그래머가 고의로 예외를 발생 시킬수 있다
~~~java
public class ExceptionEx1 {
    public static void main(String[] args) {

        try {
            throw new Exception("고의로 Exception 발생");
        } catch (Exception e) {
            System.out.println("에러 메세지: " + e.getMessage());
            e.printStackTrace();
        }
        System.out.println("프로그램이 정상 종료되었음");
    }
}
~~~


## 자바가 제공하는 예외 계층 구조

## Exception과 Error의 차이는?
- 자바에서는 실행 시(runtime) 발생할 수 있는 프로그램 오류를 Error 와 Exception 두가지로 구분
- Error 는 메모리 부족(OutOfMemoryError)이나 , StackOverflowError 와 같이 일단 발생하면 복구 할 수 없는 심각한 오류
- Exception은 발생하면 프로그래머가 이에 대한 적절한 코드를 작성해 놓음으로써 프로그램의 비정상적인 종료를 막을 수 있다.
  
## RuntimeException과 RE가 아닌 것의 차이는?

## 커스텀한 예외 만드는 방법


## 참고
- 자바의 정석 책